### DISCUSSION TITLE
Pass Configuration Variables to workflows triggered by a pull request from a fork

### DISCUSSION THREAD

**Discussion by galargh** [community/community#44322] [Actions]
### Select Topic Area

Product Feedback

### Body

The new [Configuration Variables](https://github.blog/changelog/2023-01-10-github-actions-support-for-configuration-variables-in-workflows/) feature is great! I'm baffled about their visibility scope, though.

> Variables are not passed to workflows that are triggered by a pull request from a fork. [Learn more about variables](https://docs.github.com/actions/learn-github-actions/variables).

They could/should be passed to workflows triggered by a pull request from a fork.

Why?
- they are **non-sensitive**
- they are not obfuscated in GitHub Actions which makes them visible to the public already anyway

**Comment by tbouffard** [comment]
Hi, I face the same problem today.
Dependabot doesn't have access to Configuration as well, see #44088. Everything happens as if the restrictions for the Configuration Variables were the same as for the Secrets.. However, as mentioned in the description of this "Product Feedback", they are non-sensitive.

In my case, following the examples provided by the official documentation (https://docs.github.com/en/actions/learn-github-actions/variables#using-the-vars-context-to-access-configuration-variable-values), we use a Configuration Variable to declare the Runner used by the  workflow. Not having access to the related runner variable always generates an error for PR created from fork or by Dependabot.
This make the usage of Configuration Variables useless.

Workflow definition: https://github.com/process-analytics/bpmn-visualization-js/blob/49315fa7136776a29053b6724f867329a919a5b7/.github/workflows/dependency-review-pr.yml#L12-L15
```yaml
jobs:
  dependency-review:
    runs-on: ${{ vars.RUNNER_UBUNTU }}
```
`Error when evaluating 'runs-on' for job 'dependency-review'. .github/workflows/dependency-review-pr.yml (Line: 14, Col: 14): Unexpected type of value '', expected type: OneOf.`

**Comment by tauhid621** [comment]
The [`pull_request_target`](https://docs.github.com/en/actions/using-workflows/events-that-trigger-workflows#pull_request_target) event can be used for scenarios where variables need to be available in a forked PR. This event runs in the context of the base of the pull request, rather than in the context of the merge commit, as the `pull_request` event does.

One of the reason to not allow variables available in forked PRs `pull_request` event was because of org variables being available in the workflow run. Workflow logs do show up the variables in plain text, but only the variables which are used in the workflow. Allowing a user to view all variables (Eg: using `toJson(vars)`) did not seem correct as the variables cannot be accessed with `read` permissions on the repo.

We are still exploring this area and taking feedback on the behaviour.

**Reply by ADTC** [reply]
Can the `pull_request_target` event allow the workflow run to access secrets as well (in the forked PR)?

**Reply by ADTC** [reply]
Also, would it be possible to use `workflow_run` to trigger a second workflow which has access to secrets and variables? Perhaps a workflow for `pull_request` can just complete immediately and trigger another workflow for `workflow_run: completed` event.

**Reply by ADTC** [reply]
It looks like `pull_request_target` should not be used to build or run code from the pull request because of security flaw. Since it runs in the context of PR base, it doesn't contain the changes anyway, which makes it useless for building.

**Reply by Sheargrub** [reply]
Seconding that another look at this behavior would be appreciated. Given that [security concerns around `pull_request_target` have recently come to the forefront](https://securitylab.github.com/resources/github-actions-preventing-pwn-requests/), it is confusing and frustrating that there isn't a straightforward way to pass repository-level configurations to `pull_request`-triggered actions.

**Comment by djbrown** [comment]
I fully support the restriction on Secrets. But please make Variables available to public Pull Request Workflows!! 🙏🏾 
Variables do not contain sensitive data by definition, and thus should be publicly accessible from anywhere or any Level.
The Argument regarding Organization Variables doesn't really make sense to me, since they actually should be visible.

**Reply by djbrown** [reply]
*accessible for all scopes that already have read permission to the project 😉 
@tauhid621 did you come to a new view or are there any other reasons to justify the current behaviour?
I really hope we can improve the current situation for the sake of Open Source Collaboration via GitHub Pull Requests ❤

**Reply by webbnh** [reply]
@lppppu, given that GH workflows offer the _secrets_ capability which are distinct from _variables_, why **shouldn't** variables be "public", at least to anything like a pull request, which already has substantial access to the organization or repo?

**Comment by danepowell** [comment]
At the very least, the Variables documentation needs to be updated to mention that variables are not available to pull requests. I spent an hour reviewing docs and debugging why variables were not appearing in my PR actions only to find an obscure reference in this discussion.

**Comment by pbrezina** [comment]
This took me lots of time to figure out that the variables are not available for pull request scope. And +1 to the opinion that it makes this feature useless.

**Comment by ThiefMaster** [comment]
FFS, please fix this. What's even the point of variables instead of secrets when they are not accessible in PR runs. Or just give us a goddamn setting whether a variable should be accessible or not...

**Comment by ross** [comment]
I was recently excited to think of using org-level variables to define the versions of Python all of octoDNS's CI jobs should run against so that we can avoid ~30 PRs ever time active versions change. I went through and implemented a setup that does exactly what I wanted 

https://github.com/octodns/octodns/pull/1081#issuecomment-1751293056

and it worked perfectly for my PRs, but as soon as I went to work on a 3rd party PR it failed. i'd assumed that "variables" would be available to the forks since they're not "secrets" or at all sensitive, otherwise what's the purpose of having a separate category.  I definitely do not want the PRs to have a RW github token, so `pull_request_target` doesn't make sense on top of it having the target's code and not the PRs. 

Are there any updates/plans here. Not looking forward to unwind all 30 PRs assuming there's no way to get the value to PRs from forks.

**Reply by ThiefMaster** [reply]
With that many repos I'd seriously consider crating a repo containing the Python versions and loading them dynamically in all the other CI jobs. Not as nice as a variable, but better than having to update 30 repos...

**Reply by ross** [reply]
Before I came up with the variables option I thought about having a prerequisite job that fetched a file with yaml/json in it that turns that into an output that the main job uses as an input (in the matrix.) 

Could feed that a github raw link for main like https://raw.githubusercontent.com/octodns/octodns/main/requirements.txt. It'd work, is a bit lower weight than fetching a whole repo, but is pretty Rube Goldberg compared to the "right" solution of variables just being available.

**Reply by ross** [reply]
My work-arounds https://github.com/octodns/octodns/pull/1084, https://github.com/octodns/octodns-azure/pull/73 (this is just an example, there's ~30 more of these, and https://github.com/octodns/terraformed/pull/24.

I lose my nice one-PR does everything setup since we now have to PR octodns/octodns, merge that and then re-plan & apply terraform to get the updated data, but until variables are no longer treated as secrets I see now way around it. 

It also takes longer for CI to run since there has to be an initial job to fetch the config so that it can be used as the matrix/input for the next job.

**Comment by gtjoseph** [comment]
This issue has been frustrating me for months.  95% of my workflow runs are triggered by pull requests from forked repos and I have no way to get CI control info without having a file in the repo, which isn't safe, or making a REST call to retrieve a file from some other location, which isn't convenient.  I don't understand why there's no facility to do this.  If you've got something secret, it should be in secrets.

**Comment by jcs090218** [comment]
I have this exact same issue. :(

**Comment by justinfx** [comment]
Just spent an hour trying to figure this out as well. I have a CI pipeline that works fine when I do PRs from the same repo. But then someone submits a PR from their fork and neither the environment variables nor the repository variables expand properly and it breaks. I don't understand how to support other people submitting PRs from their forks without this feature.

**Comment by redcode** [comment]
If forks cannot inherit repository variables from the upstream repository, then repository variables are broken by design. Why do they exist?

Please fix this. Repository variables would be great if they worked the way everyone expects them to work, not the way someone who implemented them thought they should work.

**Comment by ichorid** [comment]
This is ridiculous and makes us reconsider using GitHub for alternatives, since the forks system is the reason we choose GitHub originally

**Comment by Kagameow** [comment]
Unfortunately, the current behavior of configuration variables limits the ability to perform proper checks when creating a pull request from a fork. This is certainly not an ideal situation if we consider any workflow that uses forks.

**Comment by soumyamahunt** [comment]
Is having variables as part of environment and using specific environments for jobs a good workaround for this?

**Comment by bclark8923** [comment]
Any movement on this?

**Comment by braxton-mills** [comment]
Good lord I just wasted four hours because the docs still don't mention that the Context access for "vars" isn't scoped to PRs. 

I don't understand why this restriction was made; at the very least, please update the docs.

**Comment by carlosmiei** [comment]
+ 1 who lost hours because of this... makes no sense at all

**Comment by gustavovalverde** [comment]
At least allow a member of the repository to trigger a workflow in a PR coming from a fork, so that it can inherit those permissions. Right now, not being able to read the variables at all is a blocking issue with no workaround, since [pull_request_target](https://docs.github.com/en/actions/using-workflows/events-that-trigger-workflows#pull_request_target) is not a workaround for this.

**Comment by webbnh** [comment]
I just got bit by this _again_ (my previous project adopted the practice of not using forks in part because of this problem; I've moved to a new project where we use forks, and ran straight into this when I tried to borrow workflows from my old project).

There are so many ways to improve this situation (control access to `vars` context by organization/repo settings; grant access to `vars` by user membership; or, just embrace the fact that variables are **not _secrets_**).

As things stand, we have to choose between using forks or using `vars` -- it's one or the other.  😞

**Comment by npmitche** [comment]
+1 to people who have lost an hour+  to this issue. Please fix!

**Reply by gtjoseph** [reply]
> +1 to people who have lost an hour+ to this issue. Please fix!

I wish it was only a few hours.  I've lost weeks+ trying to find and maintain workarounds for handling PRs from forked repos.

**Reply by webbnh** [reply]
> I've lost weeks+ trying to [...] maintain workarounds for handling PRs from forked repos.

@gtjoseph, do you have anything worth sharing??  😇

**Reply by gtjoseph** [reply]
Well, for this particular issue, we have a job in our "CI" repo that dumps the variables for each of our project repos to JSON and commits them into the CI repo.  In that same repo, we have an action that reads the files sends the vars to GITHUB_OUTPUT.  The jobs triggered by PRs in our project repos call that action to retrieve the vars.

Our other issue requires an even stupider workaround.  Jobs triggered by PRs from forked repos can't even manipulate labels or comments on the PR itself.  Instead we have a separate job triggered by workflow_run to add status labels and reminder comments.  Get this though...  The payload the triggered workflow gets DOESN'T HAVE THE TRIGGERING PR NUMBER.  It does have the triggering job's run name however so we have the triggering job put the PR number in its run_name and the triggered job has to parse it.

**Reply by npmitche** [reply]
My workaround that I'm using right now is I set up a new branch (I called it "main-cfg") that has a single file in it "variablename.cfg", then in my workflow.yml, I have a step:
```yml
- name: Get variable
      run: echo "MY_VAR=$(curl https://raw.githubusercontent.com/org/repo/refs/heads/main-cfg/variablename.cfg)" >> $GITHUB_ENV
```

then I can use it in the next step(s):
```yml
- name: Echo Var
      run: echo "${{ env.MY_VAR }}"
```
I set up branch protections on `main-cfg` so only admins can bypass and restrict updating/deleting it. This enables me to be able to push an update to it directly when I need to update it and not have to go through a PR process, which I would have to do if the cfg file was directly in my default branch.

Let me know if any of you have any thoughts on how I could improve this workaround or if you have a better one. 🍻

**Reply by gtjoseph** [reply]
@npmitche I'm not sure one way is any better than another.  We're all doing hacks to get around a stupid limitation.

**Reply by gtjoseph** [reply]
> Have you explored or implemented any other strategies that might reduce the need for these complex workarounds?

Like what? I've been searching since we migrated to GitHub a year and a half ago.  What good are "Actions Variables" that don't work from Actions??  If you've got a small project with only a few contributors you can grant them access to the repo and have them submit PRs from personal branches in that repo.  Jobs triggered by those PRs will have access to the variables.  Our project has been around for 25 years with almost 400 contributors in that time.  We _have_ to use forked repos.

**Comment by redcode** [comment]
It's been two years and GitHub still hasn't done anything about repository variables. It looks like they don't plan to fix them. I wonder who is responsible for all this and more importantly, what their bosses are doing. This level of blatant and repeated negligence is intolerable.

**Comment by dwoz** [comment]
This bit us too. +1 this should be fixed.

**Comment by thomass-dev** [comment]
We definitely needs the possibility to pass `vars` to workflows triggered by a pull request from a fork.
In essence they are non-sensitive, unlike secrets. This would avoid a lot of complexity and duplicated code.

**Comment by pronebird** [comment]
It's surprising that this has not been thought through at Github. It's really annoying that all PRs coming from forks are failing

**Reply by webbnh** [reply]
Yeah.  For us/me, it has forced a choice between using forks and using Variables.  That is, for organizations/repos with complex workflows, I don't use forks anymore (particularly not for updating the workflows!).  (Yes, we could hack our way around the problem, but that's like swimming upstream, and we have better things to spend our efforts on.)

**Reply by gtjoseph** [reply]
Unfortunately, for many repos that have lots of external contributors, forks are a way of life.

**Reply by webbnh** [reply]
Which means that GH Variables are _not_....

**Reply by gtjoseph** [reply]
No, it means we've all found stupid workarounds :)

**Reply by thomass-dev** [reply]
Just by re-declaring the same variables in every workflow... quiet bad.

**Comment by hasufell** [comment]
Sounds like ya'll are misusing repository variables. It's a feature that they're not inherited by a fork. Repository variables are not much different from secrets. They should not be shared, because they are managed and configured by the repository admins.

You should write your workflows in a way that they always work in absence of repository variables (e.g. assuming defaults). Otherwise you're doing it wrong.

**Reply by gustavovalverde** [reply]
And if the default ever changes, you might need to update a bunch of workflows across several repositories.

In a lot of cases the default is the variable value coming from the org, that can be modified by the variable value of the repo; or even the GitHub environment.

You also don't want defaults if you need the pipeline to fail if no variable value was provided.

So hardcoding defaults would be the “doing it wrong” approach.

**Reply by webbnh** [reply]
> Repository variables are not much different from secrets.

Which makes them almost pointless as they are.  😖 

> They should not be shared, because they are managed and configured by the repository admins.

If they are not shared, then for whom/what are they being managed?

> You should write your workflows in a way that they always work in absence of repository variables

We use them mostly for things like setting software versions.  There isn't a "default" for that: we want the fork built and tested with the same versions that the upstream repo is built/tested with, and we don't want to have to modify all the workflows every time the admins move us to the next version.

**Comment by gtjoseph** [comment]
>  It's a feature that they're not inherited by a fork.

We don't want them _inherited_ by a fork.  We (the admins) want them visible to pull request workflows triggered _from_ a fork _to_ our repos. The workflows run in the repo the variables are defined in.

> You should write your workflows in a way that they always work in absence of repository variables (e.g. assuming defaults). Otherwise you're doing it wrong.

If you write your workflows to not need variables what's the point of having variables?

**Comment by Yuri05** [comment]
+1 for this feature. Please make it happen!

**Comment by kodawah** [comment]
chiming in, what's the point of separating vars and secrets if they behave the same! Definitely secrets shouldn't be passed to workflows, but variables should be inherited by forks, it's whole point of having a CI!

also +1 on *at least* documenting the behavior about this limitation
